lightly/loss/emp_ssl_loss.py

```diff
@@ -0,0 +1,111 @@
+"""Code for EMP-SSL Loss, largely taken from https://github.com/tsb0601/EMP-SSL"""
+
+from typing import List, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+
+
+class SimilarityLoss(nn.Module):
+    """Loss representing the similiarity between the patch embeddings and the average of the patch embeddings.
+
+    Args:
+        z_list:
+            List of patch embeddings.
+    Returns:
+        Similarity loss.
+
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__()
+
+    def forward(self, z_list: List[Tensor]) -> Tensor:
+        num_patch = len(z_list)
+        z_stacked = torch.stack(z_list, dim=0)
+        z_avg = z_stacked.mean(dim=0)
+
+        z_sim = torch.tensor(0.0, device=z_stacked.device, requires_grad=True)
+        for i in range(num_patch):
+            z_sim = z_sim + F.cosine_similarity(z_stacked[i], z_avg, dim=1).mean()
+
+        z_sim = z_sim / num_patch
+
+        return -z_sim
+
+
+class TotalCodingRate(nn.Module):
+    def __init__(self, eps: float = 0.2) -> None:
+        super(TotalCodingRate, self).__init__()
+        self.eps = eps
+
+    def compute_discrimn_loss(self, W: Tensor) -> Tensor:
+        """Discriminative Loss."""
+        d, b = W.shape
+        I = torch.eye(d, device=W.device)
+        scalar = d / (b * self.eps)
+        logdet = torch.logdet(I + scalar * W.matmul(W.T))
+        return logdet / 2.0
+
+    def forward(self, X: Tensor) -> Tensor:
+        return -self.compute_discrimn_loss(X.T)
+
+
+class EMP_SSLLoss(nn.Module):
+    """Implementation of the loss from 'EMP-SSL: Towards
+    Self-Supervised Learning in One Training Epoch' [0].
+    - [0] EMP-SSL, 2023, https://arxiv.org/abs/2304.03977
+
+    Examples:
+        >>> # initialize loss function
+        >>> loss_fn = EMP_SSLLoss(num_patches=100)
+        >>> base_transform = VICRegViewTransform() # As discussed in paper
+        >>> transform_fn = MultiCropTransform(transforms=base_transform, crop_counts=100)
+        >>>
+        >>> # generate the transformed samples
+        >>> samples = transform_fn(image)
+        >>>
+        >>> # feed through encoder head
+        >>> z = torch.cat([model(s) for s in samples])
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(z)
+    """
+
+    def __init__(
+        self,
+        eps: float = 0.2,
+        num_patches: int = 100,
+        tcr_coef: float = 1.0,
```

Comment:
I think we can drop both of these parameters. `num_patches` can be dropped because we can infer the number of patches based on the tensor passed to the loss function.

`tcr_coef` can be dropped because the number of coefficients of the total loss should be 1 - number of loss terms. The total loss `loss = patch_coef * patch_loss + tcr_coef * tcr_loss` could also be written as `loss = patch_coef / tcr_coef * patch_loss + tcr_loss`, thus "capturing" the value of the `tcr_coef` in the coefficient for the patch loss. Note that the paper uses only a single coefficient as well (lambda for the patch loss):
<img width="1143" alt="Screen Shot 2024-01-05 at 12 15 42" src="https://github.com/lightly-ai/lightly/assets/43336610/c332be21-b6e5-4d0e-8fd1-d37b12354e82">

Reply:
Sounds good. I was trying to stay faithful to the code implementation, but since the lambda term can be refactored as a combination of the coefficients, We can just have a lambda term.